presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/PushDownDereferences.java

```diff
@@ -0,0 +1,482 @@
+
+            Assignments newAssignments = Assignments.builder()
+                    .putAll(pushdownExpressions)
+                    .putIdentities(sourceSymbols)
+                    .build();
+            ProjectNode source = new ProjectNode(context.getIdAllocator().getNextId(), unnestNode.getSource(), newAssignments);
+            UnnestNode result = new UnnestNode(context.getIdAllocator().getNextId(), source, relicateSymbols, unnestNode.getUnnestSymbols(), unnestNode.getOrdinalitySymbol());
+            return Result.ofPlanNode(result);
+        }
+    }
+
+    private static Symbol replaceSymbol(Map<Symbol, Symbol> symbolsMap, Symbol symbol)
+    {
+        if (symbolsMap.containsKey(symbol)) {
+            return symbolsMap.get(symbol);
+        }
+        return symbol;
+    }
+
+    private static Expression replaceDereferences(Expression expression, Map<DereferenceExpression, Symbol> replacements)
+    {
+        return ExpressionTreeRewriter.rewriteWith(new DereferenceReplacer(replacements), expression);
+    }
+
+    private static class DereferenceReplacer
+            extends ExpressionRewriter<Void>
+    {
+        private final Map<DereferenceExpression, Symbol> expressions;
+
+        DereferenceReplacer(Map<DereferenceExpression, Symbol> expressions)
+        {
+            this.expressions = requireNonNull(expressions, "expressions is null");
+        }
+
+        @Override
+        public Expression rewriteDereferenceExpression(DereferenceExpression node, Void context, ExpressionTreeRewriter<Void> treeRewriter)
+        {
+            if (expressions.containsKey(node)) {
+                return expressions.get(node).toSymbolReference();
+            }
+            return treeRewriter.defaultRewrite(node, context);
+        }
+    }
+
+    private static List<DereferenceExpression> extractDereferenceExpressions(Expression expression)
+    {
+        ImmutableList.Builder<DereferenceExpression> builder = ImmutableList.builder();
+        new DefaultExpressionTraversalVisitor<Void, ImmutableList.Builder<DereferenceExpression>>()
+        {
+            @Override
+            protected Void visitDereferenceExpression(DereferenceExpression node, ImmutableList.Builder<DereferenceExpression> context)
+            {
+                context.add(node);
+                process(node.getBase(), context);
+                return null;
+            }
+        }.process(expression, builder);
+        return builder.build();
+    }
+
+    private static Map<DereferenceExpression, Symbol> getDereferenceSymbolMap(List<Expression> expressions, Context context, Metadata metadata, SqlParser sqlParser)
+    {
+        Set<DereferenceExpression> dereferences = expressions.stream()
+                .flatMap(expression -> extractDereferenceExpressions(expression).stream())
+                .collect(toImmutableSet());
+
+        return dereferences.stream()
+                .filter(expression -> !baseExists(expression, dereferences))
```

Comment:
Tests only cover one-level deep dereferences, e.g. msg.x and msg.y. Would you add tests with more levels to test the logic of consolidating a.b.c and a.b into a.b?